app/Comment.hs

```diff
@@ -1,15 +1,49 @@
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE OverloadedStrings #-}
 module Main (main) where
 
 import RIO
+import qualified RIO.List
 import qualified RIO.Text
+import Conduit
+import qualified EnumSet as ES
+import GHC.LanguageExtensions.Type
 import GHC.SyntaxHighlighter
+import qualified Lexer as L
+import Path
+import Path.IO
 
 main :: IO ()
 main = do
-  file <- readFileUtf8 "sample/UI.hs"
-  case map snd . filter isComment <$> tokenizeHaskell file of
-    Nothing -> error "invalid token"
-    Just ts -> writeFileUtf8 "sample/UI.txt" $ RIO.Text.unlines ts
+  runConduitRes $ sourceDirectoryDeep False "subs/ghc" .| awaitForever getCommentC
+  return ()
+
+getCommentC :: MonadIO m => FilePath -> ConduitT FilePath o m ()
+getCommentC fp
+  | ".hs" `RIO.List.isSuffixOf` fp = liftIO $ getComment fp
+  | otherwise = return ()
+
+getComment :: FilePath -> IO ()
+getComment fp = do
+  file <- readFileUtf8 fp
+  let pf = defaultParserFlags { L.pExtensionFlags = ES.fromList [QuasiQuotes,TypeSynonymInstances,TemplateHaskell,MultiParamTypeClasses,FlexibleInstances,DeriveDataTypeable,ScopedTypeVariables,Cpp,OverloadedStrings] }
+  case map snd . filter isComment <$> tokenizeHaskell' pf (preprocess file) of
+    Nothing -> putStrLn (show fp ++": invalid token")
+    Just ts -> do
+      dir <- parent <$> parseRelFile fp
+      let outDir = $(mkRelDir "output") </> dir
+      ensureDir outDir
+      writeFileUtf8 ("output/"++fp) $ RIO.Text.unlines ts
+
+preprocess :: Text -> Text
+preprocess = RIO.Text.unlines . map removeCPP . RIO.Text.lines
+
+removeCPP :: Text -> Text
```

Comment:
Can this function strip the whole block of preprocessors? In general, there are many lines between `#if` and `#endif` which has no `#` symbol at first.

Reply:
(Resolved)
In this case, the treatment is for `ghc-syntax-highlighter`, which cannot parse CPP line correctly. So this is enough for now.